src/util/strencodings.h

```diff
@@ -369,4 +371,78 @@ std::string Capitalize(std::string str);
  */
 std::optional<uint64_t> ParseByteUnits(std::string_view str, ByteUnit default_multiplier);
 
+namespace util {
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+/**
+ * ""_hex is a compile-time user-defined literal returning a
+ * `std::array<std::byte>`, equivalent to ParseHex(). Variants provided:
+ *
+ * - ""_hex_v: Returns `std::vector<std::byte>`, useful for heap allocation or
+ *   variable-length serialization.
+ *
+ * - ""_hex_u8: Returns `std::array<uint8_t>`, for cases where `std::byte` is
+ *   incompatible.
+ *
+ * - ""_hex_v_u8: Returns `std::vector<uint8_t>`, combining heap allocation with
+ *   `uint8_t`.
+ *
+ * @warning It may be necessary to use _v variants when serializing, or vice
+ *   versa, because vectors are assumed to be variable-length and serialized
+ *   with a size prefix, while arrays are considered fixed length and serialized
+ *   with no prefix.
+ *
+ * @warning It may be preferable to use _v variants to save stack space when
+ *   declaring local variables if hex strings are large. Alternately variables
+ *   could be declared constexpr to avoid using stack space.
+ *
+ * @warning Avoid _u8 variants when not necessary, as the codebase migrates to
+ *   use `std::byte` instead of `unsigned char` and `uint8_t`.
+ *
+ * @note The reason ""_hex uses `std::array` is because a heap-based
+ *   `std::vector` cannot cross the compile-time/runtime barrier.
+ */
+inline namespace hex_literals {
+namespace detail {
+
+template <size_t N>
+struct Hex {
+    std::array<std::byte, N / 2> bytes{};
+    consteval Hex(const char (&hex_str)[N])
+        // 2 hex digits required per byte + implicit null terminator
+        requires(N % 2 == 1)
+    {
+        if (hex_str[N - 1]) throw "null terminator required";
+        for (std::size_t i = 0; i < bytes.size(); ++i) {
+            bytes[i] = static_cast<std::byte>(
+                (ConstevalHexDigit(hex_str[2 * i]) << 4) |
+                 ConstevalHexDigit(hex_str[2 * i + 1]));
+        }
+    }
+};
+
+} // namespace detail
+
+template <util::detail::Hex str>
+consteval auto operator""_hex() { return str.bytes; }
+
+template <util::detail::Hex str>
+consteval auto operator""_hex_u8() { return std::bit_cast<std::array<uint8_t, str.bytes.size()>>(str.bytes); }
+
+template <util::detail::Hex str>
+constexpr auto operator""_hex_v() { return std::vector<std::byte>{str.bytes.begin(), str.bytes.end()}; }
+
+template <util::detail::Hex str>
+inline auto operator""_hex_v_u8() { return std::vector<uint8_t>{UCharCast(str.bytes.data()), UCharCast(str.bytes.data() + str.bytes.size())}; }
```

Comment:
In commit "util: Add consteval ""_hex[_v][_u8] literals" (f9fed4d4e6daa2bdd715b92b6bfffdbe8202a9c8)

This could be constexpr and I think switching between consteval and constexpr just makes the definitions inconsistent and adds noise. Would suggest:

```diff
--- a/src/util/strencodings.h
+++ b/src/util/strencodings.h
@@ -431,16 +431,16 @@ struct Hex {
 } // namespace detail
 
 template <util::detail::Hex str>
-consteval auto operator""_hex() { return str.bytes; }
+constexpr auto operator""_hex() { return str.bytes; }
 
 template <util::detail::Hex str>
-consteval auto operator""_hex_u8() { return std::bit_cast<std::array<uint8_t, str.bytes.size()>>(str.bytes); }
+constexpr auto operator""_hex_u8() { return std::bit_cast<std::array<uint8_t, str.bytes.size()>>(str.bytes); }
 
 template <util::detail::Hex str>
 constexpr auto operator""_hex_v() { return std::vector<std::byte>{str.bytes.begin(), str.bytes.end()}; }
 
 template <util::detail::Hex str>
-inline auto operator""_hex_v_u8() { return std::vector<uint8_t>{UCharCast(str.bytes.data()), UCharCast(str.bytes.data() + str.bytes.size())}; }
+constexpr inline auto operator""_hex_v_u8() { return std::vector<uint8_t>{UCharCast(str.bytes.data()), UCharCast(str.bytes.data() + str.bytes.size())}; }
 
 } // inline namespace hex_literals
 } // namespace util

```

Reply:
I want to explicitly enforce compile time where possible, hence the `consteval`.

In the `""_hex_v_u8` case, `constexpr` would be a lie, as `UCharCast` being called is `inline`, not `constexpr`, and calls `reinterpret_cast` which is not supported at compile time. (This means that `UCharSpanCast` and `MakeUCharSpan` are `constexpr` in name only, except for if called on a collection that is already `unsigned char` and doesn't require `reinterpret_cast`).

Reply:
Since we're templating these on `Hex` instances, this string literal approach will produce quite a rather large number of template instantiations. I wonder if keeping these `consteval` would also better allow the compiler to make sure these instantiations don't end up in the binary?

Reply:
> makes the definitions inconsistent and adds noise

I agree that it's likely just a documentation change for the reader (unsurprisingly, my trivial godbolt experiment also showed the same compiled output), but the methods *are* a bit different, isn't it useful signal if we mark them as such?

Reply:
Since the `consteval`/`constexpr`/`inline` choices are not obvious, but have fairly solid reasoning behind them, I could add non-Doxygen comments like
```C++
// Calls non-constexpr function so can only be inline
inline auto operator""_hex_v_u8() { return std::vector<uint8_t>{UCharCast(...
``
What do you think?

Reply:
That wouldn't explain to me why the compiler seems to allow it

Reply:
@stickies-v:
> It would, in my view, make it more confusing though, by hiding that these operators are in fact consteval. I might be missing some nuance here, though?

Sorry for not addressing this before, was a bit rushed in my previous post.
My understanding is the `Hex`-instantiations are `consteval`, and the operators are *instantiated* at compile time, but the last two are commonly *executed* at runtime (with a compile time baked `Hex`-value).

Reply:
As long as `Hex` is `consteval` I'm happy. Changed the literals that could be into `constexpr` and added comment about the `inline` outlier.